src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,711 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Net;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.Serialization;
+
+// This HttpHandlerDiagnosticListener class is applicable only for .NET 4.6, and not for .NET core.
+// If you are making these changes, please test your changes manually via custom test applications.
+
+namespace System.Diagnostics
+{
+    /// <summary>
+    /// A HttpHandlerDiagnosticListener is a DiagnosticListener for .NET 4.6 where HttpClient
+    /// doesn't have a DiagnosticListener built in. This class is not used for .NET Core because
+    /// HttpClient in .NET Core already emits DiagnosticSource events. This class compensates for
+    /// that in .NET 4.6. HttpHandlerDiagnosticListener has no public constructor. To use this, 
+    /// the application just needs to call <see cref="DiagnosticListener.AllListeners" /> and
+    /// <see cref="DiagnosticListener.AllListenerObservable.Subscribe(IObserver{DiagnosticListener})"/>,
+    /// then in the <see cref="IObserver{DiagnosticListener}.OnNext(DiagnosticListener)"/> method,
+    /// when it seems the System.Net.Http.Desktop source, subscribe to it. This will trigger the
+    /// initialization of this DiagnosticListener.
+    /// </summary>
+    internal sealed class HttpHandlerDiagnosticListener : DiagnosticListener
+    {
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Predicate<string> isEnabled)
+        {
+            Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Func<string, object, object, bool> isEnabled)
+        {
+            Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+        /// <summary>
+        /// Initializes all the reflection objects it will ever need. Reflection is costly, but it's better to take
+        /// this one time performance hit than to get it multiple times later, or do it lazily and have to worry about
+        /// threading issues. If Initialize has been called before, it will not doing anything.
+        /// </summary>
+        private void Initialize()
+        {
+            lock (this)
```

Comment:
It is not clear why doing the Write outside the lock is important, and it does complicate the code.  

However thinking about it the Write() call is not useful and we should probably just get rid of it.  The problem is that we KNOW that this code happen JUST BEFORE the first SUBSCRIBE to this DiagnosticListener.   When at that point there are GUARENTEED to be no subscribers, so the Write HAS to do nothing.   

Fixing this would require remembering the exception and logging it when the subscribers come along.   I don't think this is worth it as the exception was already thrown, so we do have SOME indication that something went wrong, and it is not clear that having the DiagnosticSource tell us is that important (we can add it later if we find it is useful, but frankly this kind of thing is best effort).  

Thus I think we should just catch and swallow the exception for now (Brahmnes original code).